// image-scraper/index.js
var fs = require("fs");
var http = require("http");
var https = require("https");
var path = require("path");
var url = require("url");
var events = require("events").EventEmitter;
var util = require("util");

var cheerio = require("cheerio");

// The "Image" class.
function Image(image, address){

	var at = this.attributes = image.attribs;

	this.name = path.basename(at.src, path.extname(at.src));
	this.saveTo = path.dirname(require.main.filename) + "/";
	this.extension = path.extname(at.src);
	this.address = url.resolve(address, at.src);
	this.fromAddress = address;
}

Image.prototype.save = function(callback){

	var parsedUrl = url.parse(this.address);

	// Make a reference to the current instance.
	var ref = this;

	// Support HTTPS.
	if(parsedUrl.protocol == "https:") var protocol = https;
	else var protocol = http;

	var request = protocol.request(this.address, function(response){

		if(response.statusCode != 200){
		
			console.error("Image scraper(3): image couldn't be found.");
		}
		else{

			var imageFile = fs.createWriteStream(path.normalize(ref.saveTo + ref.name + ref.extension));

			imageFile.on("error", function(e){

				console.error("Image scraper(4): error while loading image: " + e + ".");
			});

			response.on("data", function(data){

				imageFile.write(data);
			});

			response.on("end", function(){

				imageFile.end();
				
				if(typeof(callback) == "function") callback.call(ref);
			});
		}
	});

	request.end();
	request.on("error", function(e){

		console.error(e);
	});
};

function Scraper(address){
	events.call(this);
	this.address = address;
};

// Inherit the methods of "events".
util.inherits(Scraper, events);

Scraper.prototype.scrape = function(callback){

	if(typeof(callback) == "function"){

		this.on("image", callback);
	}

	var parsedUrl = url.parse(this.address);

	// Make a reference to the current instance.
	var ref = this;

	// Support HTTPS.
	if(parsedUrl.protocol == "https:") var protocol = https;
	else var protocol = http;

	var request = protocol.request(this.address, function(response){

		if(response.statusCode != 200){
		
			console.error("Image scraper(1): web page couldn't be found.");
		}
		else{

			response.setEncoding("utf8");

			var previous = "",
			current;

			response.on("data", function(data){

				var current = previous + data;

				current.replace(/<img[\S\s]*?>/ig, function(m){

					var image = new Image(cheerio.load(m)("img")[0], ref.address);

					ref.emit("image", image);
				});

				var previous = data;
			});

			response.on("end", function(){

			});
		}
	});
	request.end();

	request.on("error", function(e){

		console.error("Image scraper(2): error while loading web page: " + e + ".");
	});
};

Scraper.prototype.download = function(callback){
	if(typeof(callback) == "function"){
		this.on("image", callback);
	}
	var parsedUrl = url.parse(this.address);
	console.log("hostname: " +parsedUrl.hostname);
	console.log("path: " + parsedUrl.path);
	console.log("href: " + parsedUrl.href);
	console.log("query: " + parsedUrl.query);
	// Make a reference to the current instance.
	var ref = this;
	// Support HTTPS.
	if(parsedUrl.protocol == "https:") var protocol = https;
	else var protocol = http;
	var options = {
		
	    hostname: parsedUrl.hostname,
	    path: parsedUrl.path,
	    headers: {
	      //'Accept':	'text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8',
		  //'Accept-Encoding': 'gzip, deflate',
		  //'Accept-Language': 'en-US,en;q=0.5',
		  //'Cache-Control': 'max-age=0',
	      //'Connection': 'keep-alive',
	      'Referer': 'http://cnc.dm5.com/m11420-p41/',
	      //'User-Agent':	'Mozilla/5.0 (Macintosh; Intel Mac OS X 10.9; rv:31.0) Gecko/20100101 Firefox/31.0'
	    },
	    
	    
  	};
	var request = protocol.request(options, function(response){

		if(response.statusCode != 200){
		
			console.error("Image scraper(3): image couldn't be found.");
		}
		else{

			var imageFile = fs.createWriteStream(path.normalize("./" + "1" + ".jpg"));

			imageFile.on("error", function(e){

				console.error("Image scraper(4): error while loading image: " + e + ".");
			});

			response.on("data", function(data){

				imageFile.write(data);
			});

			response.on("end", function(){

				imageFile.end();
				
				if(typeof(callback) == "function") callback.call(ref);
			});
		}
	});
	request.end();

	request.on("error", function(e){

		console.error("Image scraper(2): error while loading web page: " + e + ".");
	});
};

module.exports = Scraper;
